### PR TITLE
cls/rbd: make group_snap_list return snapshots in creation order

### PR DESCRIPTION
**Pull request by trociny**
To make this possible the snapshot order is stored on a snapshot
creation and the shapshot list result is sorted using this order
before returning to the caller.

This change adds a restriction on start_after parameter: now it
may be either the empty string id (to start at the beginning) or
an existing snapshot id. Otherwise -ERESTART is returned to inform
the caller that the "start_after" snapshot is invalid/removed and
the listing should be restarted.

Fixes: https://tracker.ceph.com/issues/51686
Signed-off-by: Mykola Golub <mgolub@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by idryomov — COMMENTED** [review]
Some high level questions:

- Where does being able to recall group snapshot creation order come up?

- Does listing actually need to be ordered or would being able to sort the collection of returned `GroupSnapshot` structures be enough?  Can `GroupSnapshot` be extended with `seq` field instead of storing it under a separate key thus doubling the number of omap entries?

- Better yet, why can't the group snapshot id just be changed to a monotonically increasing number?  Is it persisted anywhere other than as a key suffix and an image snapshot (namespace) suffix?

**Comment by trociny** [comment]
> Some high level questions:
> 
> * Where does being able to recall group snapshot creation order come up?

I realized I would need it for group snapshot mirroring. Also we have "group import/export" PR. I looked at it long time ago, so I am not 100% sure, but I believe it will not work as expected if the snapshot listing is not returned in creation order.

And actually I believe it is what users would expect when listing snapshots.

> * Does listing actually need to be ordered or would being able to sort the collection of returned `GroupSnapshot` structures be enough?  Can `GroupSnapshot` be extended with `seq` field instead of storing it under a separate key thus doubling the number of omap entries?

Storing `seq` in `GroupSnapshot` will probably work but then we will have to deal with compatibility issues for older librbd clients.

> * Better yet, why can't the group snapshot id just be changed to a monotonically increasing number?  Is it persisted anywhere other than as a key suffix and an image snapshot (namespace) suffix?

What could be a source of this number? I can imagine only the osd. So we would need a cls method to get the snapshot id. Or modify "create snapshot" cls method to generate the snapshot id and return it to the caller. Again, what to do with the older librbd clients?

File: src/cls/rbd/cls_rbd.cc [hunk 1]
```diff
@@ -7474,6 +7591,43 @@ int group_snap_set(cls_method_context_t hctx,
...
+      if (!vals.empty()) {
+        last_read = vals.rbegin()->first;
+      } else {
+        ceph_assert(!more);
+      }
+    } while (more);
+
+    std::string order_key = group::snap_order_key(group_snap.id);
+    bufferlist bl;
+    encode(++max_order, bl);
```
[the remark this answers is not in the document]
Comment (trociny) [reply src/cls/rbd/cls_rbd.cc:7626]: It is executed on the osd side, so the requests will be serialize (can't execute concurrently). Or may be I am missing your point?

**Comment by trociny** [comment]
Actually, I was considering a possibility when ordering is made on librbd side (either in api::group_snap_list or where it is needed). After getting the (unordered) list with `cls_client::group_snap_list` we could sort it using image snapshot ids (which are stored in `GroupSnapshot.snaps`) for comparison. It is a bit trickier though, as we can't just use e.g. `GroupSnapshot.snaps[0].snap_id`, because images may be from different pools and the group snapshots not necessary contain the same lists of images (images may be added and removed from the group during its lifetime), and we would have to make sure we are comparing image snap ids from the same pool. And in general case it can not be reliable: consider a case when image1 from pool1 is added to a group, snapshot is created, then image2 from pool2 is added and the image1 is removed, and another snapshot is created.

File: src/cls/rbd/cls_rbd.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (idryomov) [reply src/cls/rbd/cls_rbd.cc:7626]: No, I just missed that this is a class method.

**Comment by idryomov** [comment]
> > * Where does being able to recall group snapshot creation order come up?
> 
> I realized I would need it for group snapshot mirroring. Also we have "group import/export" PR. I looked at it long time ago, so I am not 100% sure, but I believe it will not work as expected if the snapshot listing is not returned in creation order.

Can you explain what would break in more detail (group snapshot mirroring case would be enough if you remember it better)?

> 
> And actually I believe it is what users would expect when listing snapshots.
> 
> > * Does listing actually need to be ordered or would being able to sort the collection of returned `GroupSnapshot` structures be enough?  Can `GroupSnapshot` be extended with `seq` field instead of storing it under a separate key thus doubling the number of omap entries?
> 
> Storing `seq` in `GroupSnapshot` will probably work but then we will have to deal with compatibility issues for older librbd clients.

OK, so you are containing this on the OSD side deliberately.  But then what guarantees that `snap_list` would always return a sorted vector?  What if a new librbd client (i.e. one that actually depends on this listing to be in creation order) is exposed to old OSDs?  Or if the OSDs are new but there are some pre-existing group snapshots that are missing the new `seq` key?  For the former case, I suppose a `require_osd_release`-style check could help, but what about the latter case where `orders[<ID>]` would just default to 0?

**Comment by idryomov** [comment]
> Actually, I was considering a possibility when ordering is made on librbd side (either in api::group_snap_list or where it is needed). After getting the (unordered) list with `cls_client::group_snap_list` we could sort it using image snapshot ids (which are stored in `GroupSnapshot.snaps`) for comparison.  It is a bit trickier though ...
>
> ... And in general case it can not be reliable: consider a case when image1 from pool1 is added to a group, snapshot is created, then image2 from pool2 is added and the image1 is removed, and another snapshot is created.

Yeah, I don't see a way to make it reliable with no OSD-side support as snapshot IDs aren't unique across pools.

**Comment by trociny** [comment]
> Can you explain what would break in more detail (group snapshot mirroring case would be enough if you remember it better)?

Well, if you want sync/copy a group as snapshot diffs you want the snapshot list ordered. Actually the current rbd-mirror group implementation may be is not a good example. Because eventually I implemented it in that way, that `ImageReplayer` is running for every image in the group (as for ordinary images), so they list image snapshots in the correct order, but before starting replaying a new snapshot it sends "sync" request to the "GroupReplayer", which is satisfied only after the requests from all images are received. So in current implementation the "GroupReplayer" may not need the ordered group snapshot list, On the other hand I have some unresolved edge case issues with the current implementation and it might change. And for "status" like commands, which return the mirror snapshot lists, you would want the list to be in sorted properly.

The PR #40808 (export/import for consistent group) is probably a better example. I reviewed it some time ago. I did not tested it but I suppose it will not work as expected if the `group_snap_list` produces snapshot in not "historic" order.

> OK, so you are containing this on the OSD side deliberately. But then what guarantees that snap_list would always return a sorted vector? What if a new librbd client (i.e. one that actually depends on this listing to be in creation order) is exposed to old OSDs? Or if the OSDs are new but there are some pre-existing group snapshots that are missing the new seq key? For the former case, I suppose a require_osd_release-style check could help, but what about the latter case where orders[<ID>] would just default to 0?

Yes, it will not work with older osds. My idea was that at the time I noticed the issue we had not had anything that would need this functionality in the tree yet, so if we merged this soon enough then when we eventually need this functionality it would be wide spread. Probably we could add a check on the client side (e.g. supported osd release?). Alternatively we could add `group_snap_list2` that would either return the ordered list or fail (if the osd does not support it or cannot determine the order because the snapshots were created on an older osd). But again, not sure we need to bother with this.

**Pull request closed without merging**